File: vscode-binary/src/diagnostics/strict-error.ts
import { ERROR_CODES } from "@drxporter/shared";

export class StrictError extends Error {
  code: string;
  source: string;

  constructor(code: string, message: string, source: string) {
    super(message);
    this.name = "StrictError";
    this.code = code;
    this.source = source;
  }
}

export function throwStrict(code: string, message: string, source: string): never {
  throw new StrictError(code, message, source);
}

export function assertStrict(condition: boolean, code: string, message: string, source: string): void {
  if (!condition) {
    throwStrict(code, message, source);
  }
}

export function assertUuid(uuid: string | undefined | null, context: string): void {
  assertStrict(!!uuid && uuid.length > 0, ERROR_CODES.MISSING_UUID, `Missing UUID: ${context}`, "strict-error");
}

export function assertNoDuplicate(uuid: string, seen: Set<string>, context: string): void {
  assertStrict(!seen.has(uuid), ERROR_CODES.DUPLICATE_UUID, `Duplicate UUID: ${uuid} in ${context}`, "strict-error");
}

File: roblox-plugin/src/plugin/diagnostics/strict-error.lua
local ErrorCodes = require(script.Parent["error-codes"])
local Logger = require(script.Parent.Parent.logger)

local StrictError = {}

function StrictError.throw(code, message, source)
	local entry = {
		code = code,
		severity = "fatal",
		message = message,
		source = source or "unknown",
	}

	Logger.fatal(string.format("[%s] %s (source: %s)", code, message, entry.source))
end

function StrictError.assert(condition, code, message, source)
	if not condition then
		StrictError.throw(code, message, source)
	end
end

function StrictError.assertUuid(uuid, instanceName)
	StrictError.assert(
		uuid and #uuid > 0,
		ErrorCodes.MISSING_UUID,
		"Missing drxporter-uuid on " .. (instanceName or "unknown"),
		"strict-error"
	)
end

function StrictError.assertNoDuplicate(uuid, existingMap)
	if existingMap[uuid] then
		StrictError.throw(
			ErrorCodes.DUPLICATE_UUID,
			"Duplicate UUID detected: " .. uuid,
			"strict-error"
		)
	end
end

return StrictError
